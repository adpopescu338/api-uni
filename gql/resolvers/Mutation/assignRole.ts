import { Context } from 'libs/types';
import { MutationAssignRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';

export const assignRole = async (
  _: unknown,
  { userId, roleId }: MutationAssignRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const result = await prisma.user.update({
    where: {
      id: userId,
    },
    data: {
      roles: {
        connect: {
          id: roleId,
        },
      },
    },
    include: {
      roles: {
        include: {
          permissions: true,
        },
      },
    },
  });

  return result;
};
