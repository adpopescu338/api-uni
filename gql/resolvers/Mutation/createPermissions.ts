import { Context } from 'libs/types';
import { MutationCreatePermissionsArgs } from 'libs/types/generated';

export const createPermissions = async (
  _: unknown,
  { permissions }: MutationCreatePermissionsArgs,
  { prisma }: Context,
) => {
  const createdPermissions = await prisma.permission.createMany({
    data: permissions,
  });

  return createdPermissions;
};
