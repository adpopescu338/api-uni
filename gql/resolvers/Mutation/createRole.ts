import { Context } from 'libs/types';
import { MutationCreateRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';

export const createRole = async (
  _: unknown,
  { name, description, permissionIds }: MutationCreateRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const role = await prisma.userDefinedRole.create({
    data: {
      name,
      description,
      ...(permissionIds.length && {
        permissions: {
          connect: permissionIds.map((id) => ({ id })),
        },
      }),
    },
  });

  return role;
};
