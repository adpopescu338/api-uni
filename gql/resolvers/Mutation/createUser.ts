import { Context } from 'libs/types';
import { MutationCreateUserArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';
import { getHashedPassword } from '../../../libs/getHashedPassword';

export const createUser = async (
  _: unknown,
  { input }: MutationCreateUserArgs,
  { prisma, user, cacheManager }: Context,
) => {
  throwIfNotAdmin(user);

  const password = await getHashedPassword(input.password);

  const newUser = await prisma.user.create({
    data: {
      ...input,
      password,
      isSysAdmin: input.isSysAdmin || false,
    },
  });

  cacheManager.del(cacheManager.cacheKeys.usersCount);

  return {
    ...newUser,
    roles: [],
  };
};
