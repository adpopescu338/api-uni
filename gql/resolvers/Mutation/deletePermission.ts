import { Context } from 'libs/types';
import { MutationDeletePermissionArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from '../../../libs/throwIfNotAdmin';
import { GraphQLError } from 'graphql/error/GraphQLError';

export const deletePermission = async (
  _: unknown,
  { id }: MutationDeletePermissionArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const permission = await prisma.permission.delete({
    where: {
      id,
    },
  });

  if (!permission) {
    throw new GraphQLError('Permission not found', {
      extensions: {
        code: 'PERMISSION_NOT_FOUND',
      },
    });
  }

  return { success: true };
};
