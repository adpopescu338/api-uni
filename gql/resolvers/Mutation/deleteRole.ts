import { Context } from 'libs/types';
import { MutationDeleteRoleArgs } from 'libs/types/generated';
import { throwIfNotAdmin } from 'libs/throwIfNotAdmin';
import { GraphQLError } from 'graphql';

export const deleteRole = async (
  _: unknown,
  { id }: MutationDeleteRoleArgs,
  { prisma, user }: Context,
) => {
  throwIfNotAdmin(user);

  const result = await prisma.userDefinedRole.delete({
    where: {
      id,
    },
  });

  if (!result) {
    throw new GraphQLError('Role not found', {
      extensions: {
        code: 'ROLE_NOT_FOUND',
      },
    });
  }

  return { success: true };
};
