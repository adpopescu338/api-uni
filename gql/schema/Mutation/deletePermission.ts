import { gql } from 'graphql-tag';

export const deletePermission = gql`
  type Mutation {
    """
    Only sysadmins can delete permissions
    """
    deletePermission(id: ID!): VoidApiResponse!
  }
`;
