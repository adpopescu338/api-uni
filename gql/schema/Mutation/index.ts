import { logout } from './logout';
import { login } from './login';

import { createRole } from './createRole';
import { updateRole } from './updateRole';
import { deleteRole } from './deleteRole';
import { assignRole } from './assignRole';

import { createPermissions } from './createPermissions';
import { updatePermissions } from './updatePermissions';
import { deletePermission } from './deletePermission';

import { createUser } from './createUser';
import { updateUser } from './updateUser';

export const Mutation = [
  login,
  logout,

  createRole,
  updateRole,
  deleteRole,
  assignRole,

  createPermissions,
  updatePermissions,
  deletePermission,

  createUser,
  updateUser,
];
