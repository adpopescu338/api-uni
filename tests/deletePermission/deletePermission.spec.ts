import { describe, it, expect, jest } from '@jest/globals';
import { deletePermission } from '../../gql/resolvers/Mutation/deletePermission';

const ctx = {
  prisma: {
    permission: {
      delete: jest.fn(),
    },
  },
  user: {
    isSysAdmin: true,
  },
} as any;

describe('deletePermission', () => {
  it('Should be defined', () => {
    expect(deletePermission).toBeDefined();
  });

  it('Should throw an error if user is not system admin', async () => {
    await expect(
      deletePermission(undefined, { id: '' }, { user: { isSysAdmin: false } } as any),
    ).rejects.toThrowError();
  });

  it('Should throw an error if permission is not found', async () => {
    await expect(deletePermission(undefined, { id: 'id' }, ctx)).rejects.toThrowError();
  });

  it('Should delete the permission and return success: true', async () => {
    ctx.prisma.permission.delete.mockResolvedValue({ id: 'id' } as never);

    const result = await deletePermission(undefined, { id: 'id' }, ctx);
    const { lastCall } = ctx.prisma.permission.delete.mock;

    expect(result).toEqual({ success: true });
    expect(lastCall).toMatchSnapshot();
  });
});
