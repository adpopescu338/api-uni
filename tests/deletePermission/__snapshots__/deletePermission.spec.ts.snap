// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deletePermission Should delete the permission and return success: true 1`] = `
[
  {
    "where": {
      "id": "id",
    },
  },
]
`;
